pkg/collector/corechecks/gpu/nvmlmetrics/collector.go

```diff
@@ -0,0 +1,153 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux
+
+// Package nvmlmetrics holds the logic to collect metrics from the NVIDIA Management Library (NVML).
+// The main entry point is the Collector struct, which is responsible for collecting metrics from the NVML library.
+// There are multiple subsystems that can be used to collect metrics from NVML: GPM, FieldInfo functions, and general
+// API functions. The collection of metrics is done by the subsystems, implementing the subsystemCollector interface.
+package nvmlmetrics
+
+import (
+	"fmt"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"github.com/hashicorp/go-multierror"
+)
+
+// Collector is the main struct responsible for collecting metrics from the NVML library.
+type Collector struct {
+	lib        nvml.Interface
+	collectors []subsystemCollector
+	devices    []nvml.Device
+}
+
+// Metric represents a single metric collected from the NVML library.
+type Metric struct {
+	Name  string   // Name holds the name of the metric.
+	Value float64  // Value holds the value of the metric.
+	Tags  []string // Tags holds the tags associated with the metric.
+}
+
+// subsystemFactory is a function that creates a new subsystemCollector. lib is the NVML
+// library interface, and devices the list of devices that are present in the system in case
+// the subsystem needs to preallocate structures
+type subsystemFactory func(lib nvml.Interface, devices []nvml.Device) (subsystemCollector, error)
+
+// subsystemCollector defines a collector that gets metric from a specific NVML subsystem
+type subsystemCollector interface {
```

Comment:
also here, consider using the same interface (collectors support composition pattern, there are existing code examples in our repo with parent collector calling the children collectors)

Reply:
It's not exactly the same interface, as each subsystem grabs the metric from specific devices. But let me do a refactor to have each collector associated to only once device, as I've seen that simplifies code down the line a bit